### DISCUSSION TITLE
Brainstorm of features/modifications to address

### DISCUSSION THREAD

**Discussion by gabrielasd** [gabrielasd/pyeom#61] [Ideas]
Seems to me the most important issues right now are the following:
Document package: issue #58 
Addition of new-ich features: issue #54 (particle-hole ERPA is still in process of validation; after we are sure it is correct, particle-particle and hole-hole ERPA should be supported)
Improve eigensolver module: issue #53 (Maybe this task need to take priority over issue #54)

However, I would like to get some feedback on the following potential changes, none of which is a priority, but I think at least the 1st wouldn't take too much time to do. They are mostly related to the input/output interface of the code:
1- Format of one and two-electron integrals: 
The classes for the EOM methods, assume that the two-electron integrals are given in the format 
`<p q|| r s>`, however for ERPA (which is being implemented as a method of the class), we are using `<p q| r s>`. So in order to make the inputs consistent, I was thinking that we should handle the asymmetrization of the two-electron integrals internally in the EOM class.

2- Make the RDMs an optional keyword that would default to a single-slater determinant RDM:
Michael has already added a function that would support making this change [see tools.py](https://github.com/gabrielasd/eomee/blob/4d38806efd54364fb8ba1d5931c4024357496f41/eomee/tools.py#L151). But maybe this is not a change that should be made at the API level but rather as part of a wrapper?

3- Overal input/output interface:
This was something I had to do for the CAS-741 course, so there is already a way of doing it in the branch cas741 (there the arguments were passed through an input file). However, I was considering the alternative of using the command line (using argparse). This change would involve the modules:
main.py, load.py, output.py, integrlas.py and density.py 
that are currently in the branch `cas741_erpa` in case it is merged to master.
Also related to the interface of the package, we could have a wrapper function for the different EOM methods. @msricher I think you suggested doing something like:
obj = eom(params..., type="STRING")
where STRING would designate what EOM equation is being used (e.g. EOMIPDoubleCommutator for the electron-removal EOM with double commutator, or EOMExcCIS for particle-hole EOM equivalent to CIS)?

Please, @msricher (and @PaulWAyers) let me know what you think

**Comment by PaulWAyers** [comment]
1. I'd use the same format that `gbasis` uses. It would be nice to also support directly an FCIDUMP.
2. I think having the RDMs default to Slater determinant is good.
3. I like the idea of using a command line. We've done a bit of this with `Procrustes` and `ChemTools` so Farnaz and Fanwang could be helpful. I'm not totally sure how to specify the method; there are several different choices but perhaps rather than long strings we can think of something more compact. It could be that "IP" "EA" "EX" could be combined with "1" and "2" (number of operators) and "single" or "double" (number of commutators) or something like that. I'm not really sure, honestly. Just having a catalogue of named keywords is fine too.

**Comment by smiths** [comment]
Please ignore my comment if it isn't relevant, but if you are using strings inside your code to represent a finite set of choices, you might want to consider switching to enums.  Your discussion is focusing on the strings you read from the external world, which have to be strings.  What I can't tell is whether you are keeping the string representation internally.  If you are, this can cause potential coding problems with handling different cases by comparing strings.  The compiler (or interpreter) can also give more meaningful messages when using an enum.  Enums are also nicer in dictionaries.

Best wishes.  :smile:

**Reply by PaulWAyers** [reply]
Seems like that might be a good idea:
https://docs.python.org/3/library/enum.html
